### ISSUE TITLE
Support multiple MC versions and/or multiple APIs

### ISSUE THREAD

**Issue by ItzDerock** [ItzDerock/simplevoicechat-music#1]
Possibly utilize https://github.com/henkelmax/admiral as suggested by Simple Voice Chat developer, though this is limited to forge, fabric, and neoforge (no bukkit).

**Comment by Goofert42** [comment]
Here is a build for 1.20.1. If you want me to port the mod to more versions please DM me on discord.

[Here is the ported mod](https://kappa.lol/NAy7U)

**Comment by SofieBrink** [comment]
@JefferyPlayz Do you happen to have a build for 1.21.1 aswell? Running into the following issues when attempting to use the 1.21 version on 1.21.1. Would love to use this mod if you had an updated version available!


```
[21:24:29] [lava-daemon-pool-playback-1-thread-1/ERROR]: Error in playback of VZ-oGLluGAc
com.sedmelluq.discord.lavaplayer.tools.FriendlyException: Something broke when playing the track.
        at com.sedmelluq.discord.lavaplayer.tools.ExceptionTools.wrapUnfriendlyExceptions(ExceptionTools.java:45) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:121) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.lambda$executeTrack$2(DefaultAudioPlayerManager.java:339) ~[simple-voice-chat-music-0.2.0.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.lang.RuntimeException: Not success status code: 403
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.validateStatusCode(PersistentHttpStream.java:88) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.attemptConnect(PersistentHttpStream.java:124) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.connect(PersistentHttpStream.java:106) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.internalRead(PersistentHttpStream.java:158) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.read(PersistentHttpStream.java:174) ~[simple-voice-chat-music-0.2.0.jar:?]
        at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:295) ~[?:?]
        at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275) ~[?:?]
        at com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaEbmlReader.readEbmlInteger(MatroskaEbmlReader.java:39) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaFileReader.readNextElement(MatroskaFileReader.java:47) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaStreamingFile.readFile(MatroskaStreamingFile.java:94) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaAudioTrack.loadMatroskaFile(MatroskaAudioTrack.java:52) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaAudioTrack.process(MatroskaAudioTrack.java:35) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.track.DelegatedAudioTrack.processDelegate(DelegatedAudioTrack.java:25) ~[simple-voice-chat-music-0.2.0.jar:?]
        at dev.lavalink.youtube.track.YoutubeAudioTrack.processStatic(YoutubeAudioTrack.java:148) ~[simple-voice-chat-music-0.2.0.jar:?]
        at dev.lavalink.youtube.track.YoutubeAudioTrack.processWithClient(YoutubeAudioTrack.java:116) ~[simple-voice-chat-music-0.2.0.jar:?]
        at dev.lavalink.youtube.track.YoutubeAudioTrack.process(YoutubeAudioTrack.java:71) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:109) ~[simple-voice-chat-music-0.2.0.jar:?]
        ... 4 more
[21:24:29] [lava-daemon-pool-playback-1-thread-1/ERROR]: Failed to play Good Luck, Babe! due to error: Something broke when playing the track.
[21:24:29] [lava-daemon-pool-playback-1-thread-1/ERROR]: Handler of event com.sedmelluq.discord.lavaplayer.player.event.TrackExceptionEvent@259847a8 threw an exception.
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@3e2cfa3c[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@76a1f110[Wrapped task = dev.derock.svcmusic.audio.GroupManager$$Lambda/0x000002804631f4a0@696554c2]] rejected from java.util.concurrent.ScheduledThreadPoolExecutor@3d104af8[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 287]
        at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2081) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:841) ~[?:?]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:340) ~[?:?]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:562) ~[?:?]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.execute(ScheduledThreadPoolExecutor.java:705) ~[?:?]
        at java.base/java.util.concurrent.Executors$DelegatedExecutorService.execute(Executors.java:754) ~[?:?]
        at dev.derock.svcmusic.audio.GroupManager.nextTrack(GroupManager.java:179) ~[simple-voice-chat-music-0.2.0.jar:?]
        at dev.derock.svcmusic.audio.TrackScheduler.onTrackException(TrackScheduler.java:43) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter.onEvent(AudioEventAdapter.java:76) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayer.dispatchEvent(DefaultAudioPlayer.java:357) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayer.onTrackException(DefaultAudioPlayer.java:367) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:125) ~[simple-voice-chat-music-0.2.0.jar:?]
        at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.lambda$executeTrack$2(DefaultAudioPlayerManager.java:339) ~[simple-voice-chat-music-0.2.0.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]

```

**Comment by Goofert42** [comment]
> @JefferyPlayz Do you happen to have a build for 1.21.1 aswell? Running into the following issues when attempting to use the 1.21 version on 1.21.1. Would love to use this mod if you had an updated version available!
> 
> 
> ```
> [21:24:29] [lava-daemon-pool-playback-1-thread-1/ERROR]: Error in playback of VZ-oGLluGAc
> com.sedmelluq.discord.lavaplayer.tools.FriendlyException: Something broke when playing the track.
>         at com.sedmelluq.discord.lavaplayer.tools.ExceptionTools.wrapUnfriendlyExceptions(ExceptionTools.java:45) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:121) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.lambda$executeTrack$2(DefaultAudioPlayerManager.java:339) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
>         at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
> Caused by: java.lang.RuntimeException: Not success status code: 403
>         at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.validateStatusCode(PersistentHttpStream.java:88) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.attemptConnect(PersistentHttpStream.java:124) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.connect(PersistentHttpStream.java:106) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.internalRead(PersistentHttpStream.java:158) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.tools.io.PersistentHttpStream.read(PersistentHttpStream.java:174) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at java.base/java.io.DataInputStream.readUnsignedByte(DataInputStream.java:295) ~[?:?]
>         at java.base/java.io.DataInputStream.readByte(DataInputStream.java:275) ~[?:?]
>         at com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaEbmlReader.readEbmlInteger(MatroskaEbmlReader.java:39) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaFileReader.readNextElement(MatroskaFileReader.java:47) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaStreamingFile.readFile(MatroskaStreamingFile.java:94) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaAudioTrack.loadMatroskaFile(MatroskaAudioTrack.java:52) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.container.matroska.MatroskaAudioTrack.process(MatroskaAudioTrack.java:35) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.track.DelegatedAudioTrack.processDelegate(DelegatedAudioTrack.java:25) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at dev.lavalink.youtube.track.YoutubeAudioTrack.processStatic(YoutubeAudioTrack.java:148) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at dev.lavalink.youtube.track.YoutubeAudioTrack.processWithClient(YoutubeAudioTrack.java:116) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at dev.lavalink.youtube.track.YoutubeAudioTrack.process(YoutubeAudioTrack.java:71) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:109) ~[simple-voice-chat-music-0.2.0.jar:?]
>         ... 4 more
> [21:24:29] [lava-daemon-pool-playback-1-thread-1/ERROR]: Failed to play Good Luck, Babe! due to error: Something broke when playing the track.
> [21:24:29] [lava-daemon-pool-playback-1-thread-1/ERROR]: Handler of event com.sedmelluq.discord.lavaplayer.player.event.TrackExceptionEvent@259847a8 threw an exception.
> java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@3e2cfa3c[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@76a1f110[Wrapped task = dev.derock.svcmusic.audio.GroupManager$$Lambda/0x000002804631f4a0@696554c2]] rejected from java.util.concurrent.ScheduledThreadPoolExecutor@3d104af8[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 287]
>         at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2081) ~[?:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:841) ~[?:?]
>         at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:340) ~[?:?]
>         at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:562) ~[?:?]
>         at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.execute(ScheduledThreadPoolExecutor.java:705) ~[?:?]
>         at java.base/java.util.concurrent.Executors$DelegatedExecutorService.execute(Executors.java:754) ~[?:?]
>         at dev.derock.svcmusic.audio.GroupManager.nextTrack(GroupManager.java:179) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at dev.derock.svcmusic.audio.TrackScheduler.onTrackException(TrackScheduler.java:43) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter.onEvent(AudioEventAdapter.java:76) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayer.dispatchEvent(DefaultAudioPlayer.java:357) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayer.onTrackException(DefaultAudioPlayer.java:367) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor.execute(LocalAudioTrackExecutor.java:125) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager.lambda$executeTrack$2(DefaultAudioPlayerManager.java:339) ~[simple-voice-chat-music-0.2.0.jar:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
>         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
>         at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
> 
> ```

Hi! I can attempt to port it for you. Will get back to you ASAP.

**Comment by SofieBrink** [comment]
> Hi! I can attempt to port it for you. Will get back to you ASAP.

That'd be awesome, I tried to myself earlier without luck so i'm hopeful you might have more experience in this field then i do 😄 
Was my first time attempting to compile a minecraft mod.

**Comment by ItzDerock** [comment]
@SofieBrink I believe the error you are seeing is not related to the Minecraft mod version, but rather related to #4. 

TLDR is that YouTube implemented changes targeted towards preventing apps from streaming audio from their platform, impacting this plugin among other things. A fix is on it's way, but I have no ETA I can provide atm.

**Comment by SofieBrink** [comment]
I don’t believe it is, the errors i’m seeing don’t say anything about failing to authenticate or having to log in, and they also occur when attempting to play from soundcloud.

**Comment by ItzDerock** [comment]
> Caused by: java.lang.RuntimeException: Not success status code: 403

A [403 status code](<https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403>) signifies that the provider (in this case YouTube) responded with HTTP 403 Forbidden code. While the "Sign in to confirm your not a robot" message is more common and more indicative of the issue, I've had reports of people receiving a 403 instead while using the native 1.20 version. 

Lastly, the mod was able to start up without issues and based on the stack trace given, the error occurs in the [lavaplayer](<https://github.com/lavalink-devs/lavaplayer>) library (`at com.sedmelluq.discord.lavaplayer.*`). Lavaplayer isn't tied to Minecraft at all -- it's a completely independent library so I'm not sure how a minor 1.20 -> 1.20.1 update could break it.

To confirm this, if you have the capacity, try loading the mod on your server with 1.20 and see if it works or if the same issue persists. This would be the best way to check if it is or isn't a version issue.

As for Soundcloud failing, not entirely sure why that might be the case. Would have to investigate that separately. 

I'm currently very busy with school and other irl stuff so it may be a bit before I'm able to upload a fix.

**Comment by SofieBrink** [comment]
Sure i'll see what happens in the supported 1.21 version. No worries about being busy, i know the feeling all to well.
Edit: the mod indeed doesn't work on 1.21 either, not with Youtube, Youtube Music or Soundcloud

**Comment by JefferyPlayz** [comment]
> Sure i'll see what happens in the supported 1.21 version. No worries about being busy, i know the feeling all to well.
> Edit: the mod indeed doesn't work on 1.21 either, not with Youtube, Youtube Music or Soundcloud

Hiya. Not sure if you misread my original post with the ported version or if you are referring to another version. But the version I ported before was for 1.20.1 not 1.21.

**Comment by SofieBrink** [comment]
> Hiya. Not sure if you misread my original post with the ported version or if you are referring to another version. But the version I ported before was for 1.20.1 not 1.21.

Guess this has gone a bit off topic, That message was me testing the official release of the mod on the officially supported version (1.21) which didn't work either, which seems to confirm that the issue is not with me using an incorrect version of the mod but with lavaplayer itself
